tests/difference/core/quint_model/ccv_statemachine.qnt

```diff
@@ -0,0 +1,506 @@
+        // when the provider has just entered a validator set into a block...
+        if (ValUpdatePrecondition and currentState.providerState.providerValidatorSetChangedInThisBlock) {
+            val providerValSetInCurBlock = providerValidatorHistory.head()
+            // ... for each consumer that is running then ...
+            runningConsumers.forall(
+                // ...the validator set is in a sent packet
+                consumer => currentState.providerState.sentVscPacketsToConsumer.get(consumer).toSet().exists(
+                    packet => packet.validatorSet == providerValSetInCurBlock
+                )
+            )
+        } else {
+            true
+        }
+
+    // Every consumer chain receives the same sequence of 
+    // ValidatorSetChangePackets in the same order.
+    // NOTE: since not all consumer chains are running all the time,
+    // we need a slightly weaker invariant:
+    // For consumer chains c1, c2, if both c1 and c2 received a packet p1 sent at t1 and a packet p2 sent at t2,
+    // then both have received ALL packets that were sent between t1 and t2.
+    val SameVscPacketsInv =
+        runningConsumers.forall(
+            consumer1 => runningConsumers.forall(
+                consumer2 => {
+                    val packets1 = currentState.consumerStates.get(consumer1).receivedVscPackets
+                    val packets2 = currentState.consumerStates.get(consumer2).receivedVscPackets
+                    val commonPackets = packets1.toSet().intersect(packets2.toSet())
+                    if (commonPackets.size() == 0) {
+                        true // they don't share any packets, so nothing to check
+                    } else {
+                        val newestCommonPacket = commonPackets.maxBy(packet => packet.sendingTime)
+                        val oldestCommonPacket = commonPackets.minBy(packet => packet.sendingTime)
+                        // get all packets sent between the oldest and newest common packet
+                        val packetsBetween1 = packets1.select(
+                            packet => packet.sendingTime >= oldestCommonPacket.sendingTime and packet.sendingTime <= newestCommonPacket.sendingTime
+                        )
+                        val packetsBetween2 = packets2.select(
+                            packet => packet.sendingTime >= oldestCommonPacket.sendingTime and packet.sendingTime <= newestCommonPacket.sendingTime
+                        )
+                        // these should be the same on both chains
+                        packetsBetween1 == packetsBetween2
+                    }
+                }
+            )
+        )
+
+    // For every ValidatorSetChangePacket received by a consumer chain at 
+    // time t, a MaturedVscPacket is sent back to the provider in the first block 
+    // with a timestamp >= t + UnbondingPeriod
+    // NOTE: because we remove the maturationTimes entry when we send the packets,
```

Comment:
Thanks, that was a really useful comment 👍
Otherwise, someone might wonder how this invariant is not violated.